cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactRepository.java

```diff
@@ -50,21 +55,34 @@
  * RemoteArtifactRepository provides a remote implementation of ArtifactRepository
  */
 public class RemoteArtifactRepository implements ArtifactRepository {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteArtifactRepository.class);
+
   private final ArtifactRepositoryReader artifactRepositoryReader;
   private final ArtifactClassLoaderFactory artifactClassLoaderFactory;
+  private final Optional<ArtifactLocalizerClient> artifactLocalizerClient;
 
   @VisibleForTesting
   @Inject
   public RemoteArtifactRepository(CConfiguration cConf, ArtifactRepositoryReader artifactRepositoryReader,
-                                  ProgramRunnerFactory programRunnerFactory) {
+                                  ProgramRunnerFactory programRunnerFactory,
+                                  Optional<ArtifactLocalizerClient> artifactLocalizerClient) {
     this.artifactRepositoryReader = artifactRepositoryReader;
     this.artifactClassLoaderFactory = new ArtifactClassLoaderFactory(cConf, programRunnerFactory);
+    this.artifactLocalizerClient = artifactLocalizerClient;
   }
 
   @Override
   public CloseableClassLoader createArtifactClassLoader(ArtifactDescriptor artifactDescriptor,
                                                         EntityImpersonator entityImpersonator) throws IOException {
-    return artifactClassLoaderFactory.createClassLoader(artifactDescriptor.getLocation(), entityImpersonator);
+    Location location;
+    if (artifactLocalizerClient.isPresent()) {
+      location = localizeArtifact(new ArtifactId(artifactDescriptor.getNamespace(),
+                                     artifactDescriptor.getArtifactId().getName(),
+                                     artifactDescriptor.getArtifactId().getVersion().getVersion()));
+    } else {
+      location = artifactDescriptor.getLocation();
+    }
+    return artifactClassLoaderFactory.createClassLoader(location, entityImpersonator);
```

Comment:
minor nit: maybe better if localizeArtifact is responsible for returning location so lines 78-84 can be moved to localizeArtifact method and this can simply be

```java
public CloseableClassLoader createArtifactClassLoader(ArtifactDescriptor artifactDescriptor,
                                                        EntityImpersonator entityImpersonator) throws IOException {
   Location location = localizeArtifact(artifactId, artifactDescriptor.getLocation());
  return artifactClassLoaderFactory.createClassLoader(location, entityImpersonator);
}
```

Reply:
Valid point. 

Restructure the code differently and looks to be cleaner. 

In short, 
created 
class xxWithLocalization extends xxx {
}

where 
class xxx {
Location getLocation() {}
}

and 
class xxWithLocalization {
Location getLocation() { 
  // localize and return local location
}